app/services/experimentation/data_export.rb

```diff
@@ -0,0 +1,195 @@
+      @max_appointments = appts.count if appts.count > max_appointments
+
+      encounters_during_experiment = encounter_dates(patient, notification_start_date, cutoff_date)
+
+      appts.each_with_index.map do |appt, index|
+        # pulling out encounters sequentially because there's no formal relationship to appointments
+        followup_date = encounters_during_experiment[index]
+        days_til_followup = nil
+        bp_at_followup = nil
+        facility = nil
+
+        if followup_date
+          days_til_followup = (appt.scheduled_date - followup_date).to_i
+          encounter = encounter_by_date(patient, followup_date)
+          bp_at_followup = encounter.class == BloodPressure
+          facility = encounter.facility
+        end
+        [appt.device_created_at.to_date, appt.scheduled_date, followup_date, days_til_followup, bp_at_followup,
+         facility&.name, facility&.facility_type, facility&.state, facility&.district, facility&.block]
+      end
+    end
+
+    def pad_appointments
+      aggregate.each do |patient_data|
+        appointments_deficit = @max_appointments - patient_data[:appointments].count
+        appointments_deficit.times do
+          patient_data[:appointments] << Array.new(10, nil)
+        end
+      end
+    end
+
+    def process_past_visits(patient)
+      end_date = notification_start_date - 1.day
+      start_date = end_date - 1.year
+      encounter_dates = encounter_dates(patient, start_date, end_date)
+      @max_past_visits = encounter_dates.count if encounter_dates.count > max_past_visits
+      encounter_dates
+    end
+
+    def pad_past_visits
+      aggregate.each do |patient_data|
+        past_visits_deficit = @max_past_visits - patient_data[:encounters].count
+        past_visits_deficit.times { patient_data[:encounters] << nil }
+      end
+    end
+
+    def encounter_dates(patient, start_date, end_date)
+      date_range = (start_date.beginning_of_day..end_date.end_of_day)
+      bps = patient.blood_pressures.where(device_created_at: date_range).pluck(:device_created_at).map(&:to_date)
+      bss = patient.blood_sugars.where(device_created_at: date_range).pluck(:device_created_at).map(&:to_date)
+      pds = patient.prescription_drugs.where(device_created_at: date_range).pluck(:device_created_at).map(&:to_date)
+      encounters = (bps + bss + pds).uniq.sort
+    end
+
+    def encounter_by_date(patient, date)
+      date_range = (date.beginning_of_day..date.end_of_day)
+      bp = patient.blood_pressures.find_by(device_created_at: date_range)
+      return bp if bp
+      bs = patient.blood_sugars.find_by(device_created_at: date_range)
+      return bs if bs
+      pd = patient.prescription_drugs.find_by(device_created_at: date_range)
+      return pd if pd
+      # this shouldn't happen. consider raising error
+    end
+
+    def headers
+      keys = aggregate.first.keys
+      keys.each_with_object([]) do |key, collection|
+        case key
+        when :communications
```

Comment:
The majority of the complexity comes from the flexible number of columns in the document and the need to pad after max length has been found.